pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -88,19 +97,36 @@ message AckRequest {
  * AckResponse is the response for acknowledging datum.
  */
 message AckResponse {
+  enum Status {
+    // StatusError indicates that the offset has not been acknowledged successfully.
+    StatusError = 0;
+    // Leaving rooms for more status.
+  }
+  // FailedAcknowledgement is the acknowledgement result for a single offset.
+  // It is used by the platform to retry the acknowledgement.
+  message FailedAcknowledgement {
+    // Required field holding the index of the offset in the original batch.
+    uint64 index_at_original_batch = 1;
+    // Required field holding the offset that this ack result corresponds to.
+    Offset offset = 2;
+    // Required field holding the status of the acknowledgement.
+    Status status = 3;
```

Comment:
(tiny comment): if the status is only used in `FailedAcknowledgement`, then the status must be some kind of error.... I'm thinking if the name `StatusError` is too general?....

Reply:
Right, there can be multiple reasons causing the error. I am currently using a single `StatusError` but leaving the door open for more detailed types. e.g. `StatusTimeoutError` indicating the offset was not acked because of request timeout, `StatusAckError` indicating the offset was not acked because the source API failed to do so.

Let me add them.

Reply:
I think the error message field below is good enough?... may be we don't really need the status

Reply:
let's do status as bool, successful, or failed. let's not try to add different statuses because that is an unbounded list.

Reply:
Since the Ack API only returns failed acknowledgements, I don't see the status being successful. I think we can remove the status for now.